v2.3 Add pgbackrest & archive_command monitoring. Fix stuck pending p…

keithf4 opened the pull request:
Please check that alerts are actually firing and going into the alertmanager.

keithf4 commented on exporter/postgres/setup_pg10.sql:
```diff
@@ -10,6 +10,29 @@ GRANT pg_monitor to ccp_monitoring;
+    -- Copy data into the table directory from the pgBackRest into command
+    COPY temp_pgbackrest_data (data)
+        FROM program
+            'pgbackrest --output=json info | tr ''\n'' '' ''' (format text);
```
This is the function provided from the pgbackrest docs in order to get the current backup status in a parsable format that postgres can read.

https://pgbackrest.org/user-guide.html#monitor

Not sure how the --config option would help there since that's just used to designate the config file being used when pgbackrest is run.

sharmay replied:
config file hold repo info, which inturn read by pgbackrest to generate info output

keithf4 replied:
That's going to make things a lot more complex for configuring alerts. Will have to think on that and discuss further.

keithf4 commented on exporter/postgres/setup_pg10.sql:
```diff
@@ -10,6 +10,29 @@ GRANT pg_monitor to ccp_monitoring;
 CREATE SCHEMA IF NOT EXISTS monitor AUTHORIZATION ccp_monitoring;
 
--- There are currently no SECURITY DEFINER functions required as long as the "pg_monitor" role is granted to ccp_monitoring.
+CREATE OR REPLACE FUNCTION monitor.pgbackrest_info() returns jsonb 
```
Permanent table, or some other way to store the data for use in the function is fine.

If the info command breaks, that causes errors in the postgres logs themselves and nothing is returned by the exporter. Would have to think about how to handle that here.

keithf4 replied:
However, if the repo is unavailable for any reason, that will cause the archive_command to fail. That's one of the other new metrics I added in, so that alert would go off.

keithf4 replied:
Updated the setup_pg11.sql file to create a table to use and updated the function to use that. Also found out that the COPY PROGRAM command doesn't actually return an error if the command it calls fails. It just copies in 0 rows. So accounting for the table being empty and throwing an exception accordingly.

Unfortunately from the Prometheus side, when the check fails, the metric completely disappears from the exporter output. This means the alert never fires. The only way around this is to set up an absence alert. And the "job" names must be known in order to create absence alerts, so we really can't have these active by default. We can just provide an example like I did here

https://github.com/CrunchyData/pgmonitor/blob/master/prometheus/crunchy-alert-rules.yml#L60.

sharmay replied:
may be you add a timestamp couple and stale info can be treated as failed `pgbackrest info` command

keithf4 replied:
That seems to over complicate things. And would just be another table that's then constantly growing and needs maintenance by the user.

I've added an example absence alert for if the backrest command itself is failing. Also, while testing this, having the function now throw an explicit exception when an error is encountered causes the `PGExporterScrapeError` alert to fire as well, indicating there's a problem with the exporter collecting metrics. So by default an alert will now be thrown when the backrest command is failing, it just won't be as clear as having an explicit absence alert for the backups.

keithf4 commented:
PGBackrest monitoring should now work with multiple config files if necessary.
Also added monitoring for sequence exhaustion as well.
Fixed links to documentation on top level readme. They are case-sensitive.

keithf4 commented:
Added pgbackrest runtime metric
Updated backrest rule examples
